cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1103,223 +1105,123 @@ Slice::CsGenerator::sequenceUnmarshalCode(const SequencePtr& seq, const string&
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag << ", "
+            << getTagFormat(seq, scope) << "))";
         out << sb;
-        out << nl << stream << ".Skip(4);";
-        string tmp = "tmpVal";
-        out << nl << seqS << ' ' << tmp << ';';
-        writeUnmarshalCode(out, seq, scope, tmp, stream);
-        if(isArray)
+        if(st->minWireSize() > 1)
         {
-            out << nl << param << " = " << tmp << ";";
-        }
-        else
-        {
-            out << nl << param << " = new " << seqS << "(" << tmp << ");";
+            out << nl << stream << ".WriteSize(" << length << " * " << st->minWireSize() << " + (" << length
+                << " > 254 ? 5 : 1));";
         }
+        writeMarshalCode(out, seq, scope, param, stream);
         out << eb;
-        out << nl << "else";
+    }
+    else if(seq->hasMetaDataWithPrefix("cs:serializable:"))
+    {
+        out << nl << "if (" << param << " != null && " << stream << ".WriteOptional(" << tag
+            << ", " << getUnqualified("Ice.OptionalFormat", scope) << ".VSize))";
         out << sb;
-        out << nl << param << " = null;";
+        out << nl << stream << ".WriteSerializable(" << param << ");";
         out << eb;
     }
+    else if(isArray)
+    {
+        out << nl << stream << ".Write" << builtinSuffixTable[builtin->kind()] << "Seq(" << tag << ", "
+            << param << ");";
+    }
+    else
+    {
+        out << nl << "if (" << param << " != null)";
+        out << sb;
+        out << nl << stream << ".Write" << builtinSuffixTable[builtin->kind()] << "Seq(" << tag << ", " << param
+            << param << ".Count, " << param << ");";
+        out << eb;
+    }
+}
+
+void
+Slice::CsGenerator::writeTaggedSequenceUnmarshalCode(Output& out,
+                                                     const SequencePtr& seq,
+                                                     const string& scope,
+                                                     const string& param,
+                                                     int tag,
+                                                     const string& stream)
+{
+    const TypePtr type = seq->type();
+    BuiltinPtr builtin = BuiltinPtr::dynamicCast(type);
+    StructPtr st = StructPtr::dynamicCast(type);
+
+    out << nl << "if (" << stream << ".ReadOptional(" << tag << ", " << getTagFormat(seq, scope) << "))";
+    out << sb;
+    if(isVariableLength(type))
+    {
+        out << nl << stream << ".Skip(4);";
+    }
+    else if(st && st->minWireSize() > 1)
+    {
+        out << nl << stream << ".SkipSize();";
+    }
+    else if(builtin && builtin->kind() != Builtin::KindByte && builtin->kind() != Builtin::KindBool)
+    {
+        out << nl << stream << ".SkipSize();";
+    }
+    out << nl;
+    writeUnmarshalCode(out, seq, scope, param, stream);
+    out << eb;
+    out << nl << "else";
+    out << sb;
+    out << nl << param << " = null;";
```

Comment:
I believe this assignment to null is unnecessary.

This code can be called to unmarshal parameters, in which case we set the parameter to null before calling the function, or it can be called for Slice data members that are null-initialized automatically.

Reply:
You are right, this was not required